addons/website/static/src/snippets/s_image_gallery/000.js

```diff
@@ -99,6 +108,110 @@ const GalleryWidget = publicWidget.Widget.extend({
+            ) {
+                showMoreButton.classList.remove("d-none");
+            }
+        }
+    },
+    /**
+     * Displays hidden gallery items
+     *
+     * @private
+     */
+    _showMore() {
+        const galleryItems = this._getSortedGalleryItems();
+
+        // Show all items and toggle visibility of expand buttons
+        const showMoreButton = this.el.querySelector("#show-more-container");
+        const hideExtraButton = this.el.querySelector("#btn-hide-extra");
+
+        if (showMoreButton && hideExtraButton) {
+            showMoreButton.classList.add("d-none");
+            hideExtraButton.classList.remove("d-none");
+
+            for (const item of galleryItems) {
+                item.classList.remove("d-none");
+            }
+        }
+    },
+    /**
+     * Hides extra gallery items
+     *
+     * @private
+     */
+    _hideExtra() {
+        const galleryItems = this._getSortedGalleryItems();
+        const maxAllowedItemCount =
+            parseInt(this.el.dataset.expandableCount) || galleryItems.length;
+
+        const showMoreButton = this.el.querySelector("#show-more-container");
+        const hideExtraButton = this.el.querySelector("#btn-hide-extra");
+
+        if (showMoreButton && hideExtraButton) {
+            // Hide items beyond max allowed count
+            galleryItems.forEach((item, index) => {
+                if (index >= maxAllowedItemCount) {
+                    item.classList.add("d-none");
+                }
+            });
+
+            // Toggle expand button visibility
+            hideExtraButton.classList.add("d-none");
+            showMoreButton.classList.remove("d-none");
+
+            // Bring the gallery into view
+            showMoreButton.scrollIntoView({
+                behavior: "smooth",
+                block: "end",
+            });
+        }
+    },
+    /**
+     * Retrieves gallery items.
+     *
+     * Handles different gallery layouts (`o_grid`, `o_masonry`, `o_col`) and
+     * sorts items according to their `data-index` attribute.
+     *
+     * @private
+     * @returns {Array} An array of sorted elements in the gallery by their
+     *                  index.
+     */
+    _getSortedGalleryItems() {
+        const snippetEl = this.el;
+        let galleryItems;
+
+        if (snippetEl.classList.contains("o_grid")) {
+            // For grid layout images are wrapped in div (.o_grid_item)
+            galleryItems = Array.from(snippetEl.querySelectorAll(".o_grid_item")).sort(
+                (a, b) =>
+                    parseInt(a.querySelector("[data-index]").dataset.index) -
+                    parseInt(b.querySelector("[data-index]").dataset.index)
+            );
+        } else {
+            // For `o_masonry` and `o_col` layouts
+            galleryItems = Array.from(snippetEl.querySelectorAll("[data-index]")).sort(
+                (a, b) => parseInt(a.dataset.index) - parseInt(b.dataset.index)
+            );
+        }
```

Comment:
I don't get this logic. why we need separate logic for grid? May be there are scope of improvement to generalize the logic.

Reply:
We need separate logic for the grid layout because in other layouts, the images themselves are the gallery items, so we sort and return them based on their index.

But in the grid layout, images are wrapped in an extra `.o_grid_item` div, which becomes the gallery item. We sort these divs by the index of the image inside them and return the `.o_grid_item` elements as the gallery items.

I'll add a comment here so it's understandable why separate logic is required for the grid layout.